File: jj/programmers/level1/weapons_of_the_knights/weapons_of_the_knights.java
// source: https://school.programmers.co.kr/learn/courses/30/lessons/136798

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> countDivisorNumber(int number) {
        List<Integer> divisorNumber = new ArrayList<>();

        for (int i = 0 ; i < number + 1; i++) {
            divisorNumber.add(0);
        }

        for (int i = 1; i < number + 1; i++) {
            int j = 1;

            while (i * j < number + 1) {
                divisorNumber.set(i * j, divisorNumber.get(i *j) + 1);
                j += 1;
            }
        }

        return divisorNumber;
    }

    public int getRequiredIron(
        List<Integer> divisorNumber,
        int limit,
        int power
    ) {
        int ironRequired = 0;

        for (int i = 1; i < divisorNumber.size(); i++) {
            if (divisorNumber.get(i) > limit) {
                ironRequired += power;
            } else {
                ironRequired += divisorNumber.get(i);
            }
        }

        return ironRequired;
    }

    public int solution(int number, int limit, int power) {
        List<Integer> divisorNumber = countDivisorNumber(number);

        return getRequiredIron(divisorNumber, limit, power);
    }
}

File: jj/programmers/level1/weapons_of_the_knights/weapons_of_the_knights.py
# source: https://school.programmers.co.kr/learn/courses/30/lessons/136798

'''
각 기사는 자신의 기사 번호의 약수 개수에 해당하는 공격력을 가진 무기 구매.
협약에 의해 공격력의 제한수치가 있어, 제한수치보다 큰 공격력을 가진 무기를 구매해야 하면 협약에서 정한 공격력을 가진 무기 구매.
'''

def solution(number, limit, power):
    answer = 0
    
    divisor_number = [0] * (number + 1)
    
    for i in range(1, number + 1):
        j = 1
        
        while i * j < number + 1:
            divisor_number[i * j] += 1
            j += 1
            
    for i in range(1, number + 1):
        if divisor_number[i] > limit:
            answer += power
        else:
            answer += divisor_number[i]
    
    return answer